rir/src/compiler/translations/pir_2_rir/allocators.h

```diff
@@ -0,0 +1,470 @@
+            verifyBB(nullptr, code->entry, f, s);
+        }
+    }
+
+    size_t operator[](Value* v) const {
+        assert(allocation.at(v) != stackSlot);
+        return allocation.at(v) - 1;
+    }
+
+    size_t slots() const {
+        unsigned max = 0;
+        for (auto a : allocation) {
+            if (a.second != stackSlot && max < a.second)
+                max = a.second;
+        }
+        return max;
+    }
+
+    bool onStack(Value* v) const { return allocation.at(v) == stackSlot; }
+
+    bool hasSlot(Value* v) const { return allocation.count(v); }
+
+    size_t getStackOffset(Instruction* instr, std::vector<bool>& used,
+                          Value* what, bool remove) const {
+
+        auto stack = sa.stackBefore(instr);
+        assert(stack.size() == used.size());
+
+        size_t offset = 0;
+        size_t usedIdx = used.size() - 1;
+        auto i = stack.rbegin();
+        while (i != stack.rend()) {
+            if (*i == what) {
+                if (remove) {
+                    used[usedIdx] = true;
+                }
+                return offset;
+            }
+            if (hasSlot(*i) && onStack(*i) && !used[usedIdx]) {
+                ++offset;
+            }
+            ++i;
+            --usedIdx;
+        }
+        assert(false && "Value wasn't found on the stack.");
+        return -1;
+    }
+
+    size_t stackPhiOffset(Instruction* instr, Phi* phi) const {
+        auto stack = sa.stackBefore(instr);
+        size_t offset = 0;
+        for (auto i = stack.rbegin(); i != stack.rend(); ++i) {
+            if (*i == phi || phi->anyArg([&](Value* v) { return *i == v; }))
+                return offset;
+            if (hasSlot(*i) && onStack(*i))
+                ++offset;
+        }
+        assert(false && "Phi wasn't found on the stack.");
+        return -1;
+    }
+
+    // Check if v is needed after argument argNumber of instr
+    bool lastUse(Instruction* instr, size_t argNumber) const {
+        assert(argNumber < instr->nargs());
+        Value* v = instr->arg(argNumber).val();
+        auto stack = sa.stackAfter(instr);
+        for (auto i = stack.begin(); i != stack.end(); ++i)
+            if (*i == v)
+                return false;
+        for (size_t i = argNumber + 1; i < instr->nargs(); ++i)
+            if (instr->arg(i).val() == v)
+                return false;
+        return true;
+    }
+};
+
+struct nameAndEnvHash {
+  public:
+    template <typename T, typename U>
+    std::size_t operator()(const std::pair<T, U>& x) const {
+        return std::hash<T>()(x.first) ^ std::hash<U>()(x.second);
+    }
+};
+
+class CachePositionAllocator {
+  public:
+    typedef size_t SlotNumber;
+
+    size_t numberOfBindings() { return uniqueNumbers.size(); }
+    SlotNumber slotFor(SEXP varName, Value* environment) {
+        std::pair<SEXP, Value*> key =
+            std::pair<SEXP, Value*>(varName, environment);
+        return uniqueNumbers.emplace(key, uniqueNumbers.size() + 1)
```

Comment:
this always overrides and never returns the same value?

Reply:
emplace does not insert if the key is already defined and returns that key

Reply:
nevermind